### PR TITLE
Add iter<M: Metric<N>>() method to Cursor

### PR DESCRIPTION
**Pull request by cmyr**
This is an ergonomics improvement; it allows a cursor be used as an Iterator,
which fits naturally into existing API and useage.

## Review Checklist

- [ ] I have responded to reviews and made changes where appropriate.
- [x] I have tested the code with `cargo test --all` / `./rust/run_all_checks`.
- [x] I have updated comments / documentation related to the changes I made.
- [x] I have rebased my PR branch onto xi-editor/master.

### CODE REVIEW

File: rust/rope/src/tree.rs [hunk 1]
```diff
@@ -790,6 +808,27 @@ impl<'a, N: NodeInfo> Cursor<'a, N> {
     }
 }
 
+/// An iterator generated by a [`Cursor`], for some [`Metric`].
+pub struct CursorIter<'c, 'a: 'c, N: 'a + NodeInfo, M: 'a + Metric<N>> {
+    cursor: &'c mut Cursor<'a, N>,
+    __metric: PhantomData<&'a M>,
```
Comment (raphlinus) [inline rust/rope/src/tree.rs]: Why two underlines? I think one is standard. I could be wrong though.

Comment (cmyr) [reply rust/rope/src/tree.rs]: No better reason than, "I've seen this elsewhere". Looking into this I guess I've only seen that on things that have to be `pub` for some reason? will fix.

**Comment by cmyr** [comment]
I looked into `DoubleEndedIterator` but my understanding is that that doesn't mean that we can go backwards; it means that we can access items from either end, converging on a middle. Could be wrong?

**Comment by raphlinus** [comment]
I think that's right. Well, we can cross that again if we ever need a reverse iterator (maybe just a new `rev_iter`).

**Pull request merged**